### PR TITLE
AOT for Fuchsia.

### PR DESCRIPTION
**Pull request by rmacnak-google**
The main complication here is the ordering of initialization. On iOS and Android, the AOT snapshot comes from a fixed location on the filesystem, so the engine has been able to initialize the VM early. In Fuchsia, we don't have access to an AOT snapshot until we receive the first app bundle, so we have to delay initializing the VM until then. The AOT snapshots should be refactored such that the VM can be initialize and the service isolate started independently of any app isolate.

### CODE REVIEW

File: content_handler/BUILD.gn [hunk 1]
```diff
@@ -54,7 +54,83 @@ executable("content_handler") {
     "//flutter/flow",
     "//flutter/glue",
     "//flutter/lib/ui",
-    "//flutter/runtime",
+    "//flutter/runtime:runtime",
```
Comment (abarth) [inline content_handler/BUILD.gn]: the ":runtime" here is redundant.

File: content_handler/BUILD.gn [hunk 2]
```diff
@@ -54,7 +54,83 @@ executable("content_handler") {
...
+
+  deps = [
+    "//application/lib/app",
+    "//apps/icu_data/lib",
+    "//apps/mozart/lib/skia:vmo",
+    "//apps/mozart/services/buffers",
+    "//apps/mozart/services/buffers/cpp",
+    "//apps/mozart/services/composition",
+    "//apps/mozart/services/input",
+    "//apps/mozart/services/views",
```
Comment (abarth) [inline content_handler/BUILD.gn]: Can we use gn variables to avoid repeating ourselves so much?

File: content_handler/runtime_holder.cc [hunk 3]
```diff
@@ -103,6 +111,57 @@ void RuntimeHolder::Init(
...
+        dlsym(dylib_handle_, "_kDartVmSnapshotInstructions"));
+    default_isolate_snapshot_data = reinterpret_cast<const uint8_t*>(
+        dlsym(dylib_handle_, "_kDartIsolateSnapshotData"));
+    default_isolate_snapshot_instr = reinterpret_cast<const uint8_t*>(
+        dlsym(dylib_handle_, "_kDartIsolateSnapshotInstructions"));
+  }
+
+  // TODO(rmacnak): We should generate the AOT vm snapshot separately from
+  // each app so we can initialize before receiving the first app bundle.
+  static std::atomic_bool first_app(true);
```
Comment (abarth) [inline content_handler/runtime_holder.cc]: There no need for atomic.  There's only one thread here.

File: runtime/BUILD.gn [hunk 4]
```diff
@@ -109,3 +109,68 @@ source_set("runtime") {
...
+    "runtime_delegate.cc",
+    "runtime_delegate.h",
+    "runtime_init.cc",
+    "runtime_init.h",
+    "start_up.cc",
+    "start_up.h",
+    "test_font_data.cc",
+    "test_font_data.h",
+    "test_font_selector.cc",
+    "test_font_selector.h",
```
Comment (abarth) [inline runtime/BUILD.gn]: Same comment about using variables to avoid repeating ourselves.

**Comment by abarth** [comment]
LGTM

File: content_handler/runtime_holder.cc (hunk 3 above)
Comment (chinmaygarde) [inline content_handler/runtime_holder.cc]: Even if there were many threads, wouldn't you need to do a compare and exchange?

**Comment by rmacnak-google** [comment]
Refactoring the GN revealed the AOT content handler was linking with both libdart_precompiled_runtime and lib_dart (the JIT), and the ordering of the deps affects which one wins. I'll untangle this knot before landing this CL.

**Comment by rmacnak-google** [comment]
Please take a look:
 - Refactored GN to avoid duplication and remove use of libdart in favor of an explicit libdart_jit or libdart_precompiled_runtime
 - Added missing language flags to the gen_snapshot invocation.

**Pull request merged**